compiler-rt/lib/builtins/cpu_model.c

```diff
@@ -1259,6 +1270,64 @@ struct {
   // As features grows new fields could be added
 } __aarch64_cpu_features __attribute__((visibility("hidden"), nocommon));
 
+#if defined(__APPLE__)
+#include <TargetConditionals.h>
+#if TARGET_OS_OSX || TARGET_OS_IPHONE
+#include <dispatch/dispatch.h>
+#include <sys/sysctl.h>
+
+static bool isKnownAndSupported(const char *name) {
+  int32_t val = 0;
+  size_t size = sizeof(val);
+  if (sysctlbyname(name, &val, &size, NULL, 0))
+    return false;
+  return val;
+}
+
+void __init_cpu_features_resolver(void) {
+  static dispatch_once_t onceToken = 0;
+  dispatch_once(&onceToken, ^{
```

Comment:
I am not familiar with blocks. Does this introduce any dependency on libcalls? Can this reuse `if (__aarch64_cpu_features.features)` used by ELF?

Reply:
The only dependencies from the block + the dispatch_once should be on things in compiler-rt and in libSystem (which we need anyway for the sysctls). I am not sure whether we can get away with doing the `if (__aarch64_cpu_features.features)` check that ELF does, since the resolvers that will be calling this will happen lazily at runtime on the Darwin implementation, and therefore potentially by separate threads (unlike ELF, which would run them at load time). This function is idempotent, but I don't think that is enough of an atomicity guarantee, especially if `__aarch64_cpu_features` grows more fields.

Reply:
Thanks for the explanation. If `__init_cpu_features_resolver` can be called concurrently, I think it is worth a comment, since the other `__init_*` functions are guaranteed to be very early in the program startup sequence and is serial.